deploy: remove generic deployment trigger controller

mfojtik opened the pull request:
Make main deployment config controller trigger rollouts on config changes and image changes.

This replaces the generictrigger controller which was calling instantiate to trigger rollouts, but we don't want to call instantiate from a controller.

This also handles the logic of initial deployment rollouts (with ICT/without ICT/...). The image change trigger is now delegated to @smarterclayton trigger controller so we only observe an image change which we treat differently from a config change.

This should be fully backward compatible with old logic.

mfojtik commented:
[test]

@kargakis @deads2k one less controller to worry about?

mfojtik commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -63,6 +63,8 @@ type DeploymentConfigController struct {
 	// queue contains deployment configs that need to be synced.
 	queue workqueue.RateLimitingInterface
+	// instantiateQueue contains deployment config that should be instantiated.
+	instantiateQueue workqueue.RateLimitingInterface
```
@kargakis i didn't want to pollute the main DC controller queue with instantiate calls, that's why I made separate queue just for instantiating (terrible word).

kargakis commented on pkg/deploy/controller/deploymentconfig/factory.go:
```diff
@@ -193,6 +227,42 @@ func (c *DeploymentConfigController) worker() {
+		return false
+	}
+
+	if _, err := c.dn.DeploymentConfigs(dc.Namespace).Instantiate(deployutil.DefaultInstantiateRequestForConfig(dc.Name)); err != nil {
```
No, please, we should stop using instantiate inside a controller. Also I was imagining that we could do everything in a single queue.

mfojtik replied:
@kargakis what you want to do instead? just update latest version?

kargakis replied:
We need to move away from the assumption that bumping latestVersion starts a deployment because that is a status field.

mfojtik commented:
@kargakis PTAL

[test]

mfojtik commented:
(this need a unit test to guard us from adding labels/annotations to the RC templates otherwise bad things will happen).

mfojtik commented:
@tnozicka for the review as well :)

kargakis commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -95,18 +95,6 @@ func init() {
 			rbac.NewRule("create", "get", "list", "watch", "update", "patch", "delete").Groups(kapiGroup).Resources("replicationcontrollers").RuleOrDie(),
```
Break down the removal of this controller into a separate commit.

kargakis commented on pkg/deploy/util/util.go:
```diff
@@ -274,6 +285,60 @@ func CopyApiEnvVarToV1EnvVar(in []api.EnvVar) []v1.EnvVar {
+		return true, "", err
+	}
+	rcTemplate := strippedRc.Spec.Template
+	// there is difference between map[string]string(nil) vs. map[string]string{}
```
of course there is :)

mfojtik replied:
just for clarification if somebody will wonder why i'm doing below that ;-)

kargakis commented on pkg/deploy/util/util.go:
```diff
@@ -274,6 +285,60 @@ func CopyApiEnvVarToV1EnvVar(in []api.EnvVar) []v1.EnvVar {
+	if err != nil {
+		return nil, err
+	}
+	delete(copy.Spec.Template.Annotations, deployapi.DeploymentVersionAnnotation)
```
Follow the upstream pattern: https://github.com/kubernetes/kubernetes/blob/ede78d9ee7706b5d5dab943e2b56f373308f578e/pkg/controller/deployment/util/deployment_util.go#L281

mfojtik replied:
looks nicer, sure

kargakis commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -181,6 +181,30 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+	// If we already have latest deployment running but the config template does not match
+	// the latest deployment assume this is a config change and trigger a new deployment if
+	// the config change trigger is enabled.
+	if deployutil.HasChangeTrigger(config) && !config.Spec.Paused {
```
Break this out in its own func

kargakis commented:
Fixes https://github.com/openshift/origin/issues/13763

mfojtik commented:
@kargakis addressed comments, but the integration tests looks bad.. seems like ICT is broken... i thought we now use generic triggers for ICT changes?

mfojtik commented:
@kargakis this got pretty complicated now :-) basically @smarterclayton controller is doing config change (to replace images)... however we have to record that as image change... you can also have DC with just ICT and without CTT... So the logic to handle this gets a bit crazier now.

mfojtik commented:
@kargakis 

```
ok      TestTriggers_MultipleICTs
ok      TestTriggers_configChange
ok      TestTriggers_imageChange
ok      TestTriggers_imageChange_nonAutomatic
ok      TestTriggers_manual
```

victory!

mfojtik commented:
re-[test]

mfojtik commented on pkg/deploy/util/util.go:
```diff
@@ -210,6 +226,53 @@ func HasImageChangeTrigger(config *deployapi.DeploymentConfig) bool {
+	config.Status.Details = new(deployapi.DeploymentDetails)
+	config.Status.Details.Causes = []deployapi.DeploymentCause{{
+		Type:         deployapi.DeploymentTriggerOnImageChange,
+		ImageTrigger: &deployapi.DeploymentCauseImageTrigger{From: api.ObjectReference{Kind: "DockerImage", Name: name}},
```
@kargakis i guess this is ok, but just to be sure, this will always record the docker image (not image stream)... to do that I will have to search the triggers and find the trigger with lastTriggeredImage matching the name... can do that but don't want to complicate this if we can avoid it. wdyt?

mfojtik commented on pkg/deploy/util/util.go:
```diff
@@ -375,13 +504,17 @@ func MakeDeploymentV1(config *deployapi.DeploymentConfig, codec runtime.Codec) (
 		podLabels[k] = v
 	}
+	// NOTE: You need to update the labelsToSkip variable in pkg/deploy/util when
+	// adding a new label here.
```
@kargakis guess this check is enough? i don't think we want to keep adding more annotations or labels but if we do, this will remind you to update the list.

mfojtik commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -181,6 +181,21 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		return fmt.Errorf("found previous inflight deployment for %s - requeuing", deployutil.LabelForDeploymentConfig(config))
 	}
+	// Process triggers and start an initial rollouts
+	configCopy, err := deployutil.DeploymentConfigDeepCopy(config)
```
@kargakis this is needed to avoid cache mutation (i believe)

mfojtik commented:
new-app flake

re-[test]

bparees commented:
@mfojtik if you're seeing the new-app flake on this PR, then it doesn't sound like it solves the problem of multiple deployments, since i'm pretty sure at this point that that flake is caused by multiple deployments.

mfojtik commented:
re-[test]

mfojtik commented:
re-[test]

mfojtik commented:
@smarterclayton @kargakis tests are green :-)

mfojtik commented:
@kargakis @tnozicka since this is 3.7, two more minor refactorings:

1) Move the cancellation of existing rollouts to its own separate function to make the handle more clean
2) Retry on conflicts in updateStatus() to avoid reconciliation

tnozicka commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -251,6 +217,147 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+			return false, true
+		}
+		// Rollout if we only have config change trigger.
+		if configTrigger && !imageTrigger {
```
Testing for `!imageTrigger` seems redundant since it is already ruled out by the code above. But generally this logic seems slightly confusing since it tests conditions twice. Could we do:
```go
if imageTrigger {
    if hasAvailableImages {
        return
    }
   return
}

if configTrigger {
    return
}

mfojtik replied:
@tnozicka i was trying to make it more explicit, looks like I failed :-)

mfojtik replied:
updated.

mfojtik commented:
re-[test]

kargakis commented:
/lint

mfojtik commented:
flake: https://github.com/openshift/origin/issues/16374

/retest

tnozicka commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -122,63 +121,37 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
-		c.recorder.Eventf(config, v1.EventTypeNormal, "DeploymentAwaitingCancellation", "Deployment of version %d awaiting cancellation of older running deployments", config.Status.LatestVersion)
-		return fmt.Errorf("found previous inflight deployment for %s - requeuing", deployutil.LabelForDeploymentConfig(config))
+	// Process triggers and start an initial rollouts
+	configCopy, err := deployutil.DeploymentConfigDeepCopy(config)
```
maybe we could do this deep copy at the start of this function and be sure we don't modify shared cache since now the only paths that wouldn't reach it are exceptional

tnozicka commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -122,63 +121,37 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+	// Process triggers and start an initial rollouts
+	configCopy, err := deployutil.DeploymentConfigDeepCopy(config)
+	if err != nil {
+		glog.Errorf("Unable to copy deployment config: %v", err)
```
shouldn't we re-queue here?

kargakis replied:
deep copy shouldn't return an error anymore

tnozicka commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -335,30 +308,108 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments []
+		copied.Status = newStatus
+		_, err := c.dn.DeploymentConfigs(copied.Namespace).UpdateStatus(copied)
+		if kapierrors.IsConflict(err) {
+			copied, err = getFreshCopy()
```
Say you have just acted on some state (DC), getting 409 you know that state has changed but you have acted on the old version. If you fetch the new config and calculate status from it you will set 
observedGenerationg to the new generation from the fresh copy in calculate status, noe the one you acted on.
Anyone observing that object will think the new state is reconciled because observerGeneration=Generation although just generation-1 has been reconciled.

kargakis replied:
Agreed. Let's not retry DC updates for now.

mfojtik replied:
agree, i will remove retrying

tnozicka commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -335,30 +308,108 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments []
+		// Retry faster on conflicts
+		var updatedDeployment *v1.ReplicationController
+		err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			rc, err := c.rcLister.ReplicationControllers(deployment.Namespace).Get(deployment.Name)
```
do we need this live get?

mfojtik replied:
removing the retry on conflict for now

tnozicka commented on pkg/apps/util/util.go:
```diff
@@ -276,6 +338,72 @@ func CopyApiEnvVarToV1EnvVar(in []api.EnvVar) []v1.EnvVar {
+	configTemplate = filterTemplateAnnotationsAndLabels(configTemplate)
+	rcCopy, err := DeploymentDeepCopyV1(rc)
+	if err != nil {
+		return true, "", err
```
deepcopy failing seems more like a panic to me

kargakis replied:
@sttts made changes so that deepcopies should not error out - don't we have that code already in Origin?

sttts replied:
it will come in with the 3.8 rebase (or @deads2k backporting it, am not through my mails yet after a week vacataion. Maybe we have it already in origin :-) )

mfojtik replied:
will rather wait for that than introducing panic here now ;)

mfojtik replied:
(also this is pre-existing, other deep copies do the same)

tnozicka commented on pkg/apps/util/util.go:
```diff
@@ -276,6 +338,72 @@ func CopyApiEnvVarToV1EnvVar(in []api.EnvVar) []v1.EnvVar {
+	return false, diff.ObjectReflectDiff(configTemplate, rcTemplate), nil
+}
+
+// HasUpdatedImages indicates that deployment configuration images were updated.
```
nit: s/that/if/

tnozicka commented on pkg/apps/util/util.go:
```diff
@@ -276,6 +338,72 @@ func CopyApiEnvVarToV1EnvVar(in []api.EnvVar) []v1.EnvVar {
+	}
+	for _, c := range dc.Spec.Template.Spec.Containers {
+		if !rcImages.Has(c.Image) {
+			return true, c.Image
```
@mfojtik this will return on first updated image

We can have more images to update, right? If yes, shouldn't we wait for all of them? If not this should probably be called just `HasUpdatedImage` without `s`

mfojtik replied:
@tnozicka problem with this function is that i need to return the image that was updated... if it is multiple images then the cause is a little bit tricky to construct because it will have to contain multiple images (so multiple causes?).

mfojtik replied:
i think i will prefer multiple causes in that case, that is why causes is a list ;-)

tnozicka commented:
/lint

mfojtik commented on pkg/apps/registry/deployconfig/strategy.go:
```diff
@@ -152,6 +152,10 @@ func (statusStrategy) PrepareForUpdate(ctx apirequest.Context, obj, old runtime.
 	newDc.Labels = oldDc.Labels
+
+	if newDc.Status.LatestVersion != oldDc.Status.LatestVersion {
+		newDc.Generation = oldDc.Generation + 1
```
@tnozicka that is spec vs. status... @kargakis thoughts?

kargakis replied:
I don't think we need this here. The hack inside the spec strategy linked by @tnozicka is in place because `oc deploy` was bumping status.latestVersion prior to splitting spec/status in DCs. Nothing (ie. the DC controller today) that updates /status should touch latestVersion.

kargakis replied:
Btw, either 3.7 or 3.8 we should be able to get rid of the hack altogether since we have `oc rollout latest` in the client. Depends on when we added the latter.

kargakis commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -335,30 +308,108 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments []
+		})
+		if err != nil {
+			c.recorder.Eventf(config, v1.EventTypeWarning, "DeploymentCancellationFailed", "Failed to cancel deployment %q superceded by version %d: %s", deployment.Name, config.Status.LatestVersion, err)
+			return nil
```
We should return this error, otherwise bad things may happen, right?

kargakis commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -500,3 +551,78 @@ func (c *DeploymentConfigController) cleanupOldDeployments(existingDeployments [
+	if config.Spec.Paused {
+		return false, false
+	}
+	imageTrigger := deployutil.HasImageChangeTrigger(config)
```
Is it possible to share code with instantiate?

mfojtik replied:
some parts if you talking about canTrigger, ok with follow up? don't want to bloat this PR

kargakis replied:
issue please

kargakis commented on pkg/cmd/server/origin/controller/config.go:
```diff
@@ -76,7 +75,6 @@ func (c *OpenshiftControllerConfig) GetControllerInitializers() (map[string]Init
 
 	ret["openshift.io/deployer"] = c.DeployerControllerConfig.RunController
 	ret["openshift.io/deploymentconfig"] = c.DeploymentConfigControllerConfig.RunController
-	ret["openshift.io/deploymenttrigger"] = c.DeploymentTriggerControllerConfig.RunController
```
Yay! 🎉

kargakis commented:
/lint

kargakis commented:
Hrm, @mfojtik you managed to break the linter! Well done! :D 
```
{"author":"kargakis","error":"open /tmp/git421202590/pkg/apps/controller/generictrigger/deploy_generictrigger_controller.go: no such file or directory","level":"error","msg":"Error handling IssueCommentEvent.","org":"openshift","plugin":"golint","pr":14910,"repo":"origin","time":"2017-09-17T20:22:37Z","url":"https://github.com/openshift/origin/pull/14910#issuecomment-330080590"}
```

mfojtik commented:
@kargakis by removing a file? ;-)

mfojtik commented:
/retest

tnozicka commented on pkg/apps/util/util.go:
```diff
@@ -276,6 +345,76 @@ func CopyApiEnvVarToV1EnvVar(in []api.EnvVar) []v1.EnvVar {
+			updatedImages = append(updatedImages, c.Image)
+		}
+	}
+	if len(updatedImages) == 0 {
```
@mfojtik maybe this function could return just the array as the bool property is already contained in its length?

mfojtik replied:
i would rather prefer to be explicit as the function has `Has*` prefix, I would expect boolean.

tnozicka replied:
I though rename was self suggesting for that case

mfojtik commented:
@tnozicka @kargakis comments addressed, will wait for the green tests.

* we can re-use HasTrigger() in instantiate but as a follow up
* disabled retries for status updates for now

mfojtik commented:
/retest

mfojtik commented:
hrm... this is now somehow timeouting the integration without any error reported...

mfojtik commented:
found the problem, we should be green now

deads2k commented:
/retest

kargakis commented:
Not sure why the extended test change is required. You didn't change any registry code.

mfojtik commented:
@kargakis i removed bumping the generation in update status strategy

kargakis commented:
@mfojtik so you removed code that you added in this PR? I can't see the status update strategy doing anything like that in master[1].

[1] 
https://github.com/openshift/origin/blob/3fa1968831125d1ef9b89171600c3d7504dcf207/pkg/apps/registry/deployconfig/strategy.go#L150-L155

mfojtik commented:
@kargakis in that case i dunno, perhaps collapsing triggers gives us less generation bumps?

kargakis commented:
Not sure. The only place we bump generation is the main strategy. I am not as worried about the generation bump though. Are you planning on landing this in 3.7? Feels late for such a big change to me.

mfojtik commented:
@kargakis still have 1w for feature freeze

mfojtik commented:
/retest

tnozicka commented:
@kargakis this fixes weird DC triggering behaviour I have noticed so I would prefer this to land for 3.7

kargakis commented:
I am fine with this. Needs a rebase.

mfojtik commented:
@kargakis @tnozicka rebased

mfojtik commented:
Tagging as bug fix since this has potential to fix weird triggering behavior we seen recently.

tnozicka commented:
Fixes https://github.com/openshift/origin/issues/16671
probably lower chances of https://github.com/openshift/origin/issues/16003

tnozicka commented:
/retest

mfojtik commented:
flakes: https://github.com/openshift/origin/issues/16673 and https://github.com/openshift/origin/issues/16672

tnozicka commented:
/lgtm

mfojtik commented:
/retest

tnozicka commented:
/retest

mfojtik commented:
@bparees @jim-minter @tnozicka so I figured out what caused the continuous rollouts for the mysql deployment config. Basically the DC generated with new-app (or I assume template) has RC template with "name":

```yaml
  template:
    metadata:
      annotations:
        openshift.io/generated-by: OpenShiftNewApp
      creationTimestamp: null
      labels:
        app: cakephp-mysql-persistent
        name: mysql
      name: mysql # <--
```

However, when we create the deployment (RC) from this template, we don't copy `config.Spec.Template.Name` into the deployment. On reconcile, when the deployment config controller runs, it sees this as a configuration diff:

```
I1005 08:25:14.891263   27502 deploymentconfig_controller.go:610] Rolling out #6 deployment for myproject/mysql caused by config change, dif
f:.
object.ObjectMeta.Name:
  a: "mysql"
  b: ""
```
(thanks God I added this diffs ;-) This means the DC controller will keep bumping the LatestVersion because it thinks there was a config change in the DC but since we never copy the "name" into the RC it will do that forever... This only affects templates that has "name" set, which is the reason why none of our deployment conformance tests captured this. Also this was broken for a long time, so I dunno how the old config change trigger controller managed this....

mfojtik commented:
actually copying the name/etc might not be a good solution here because it will cause all existing DC with latest deployment not having them to trigger which might lead to a mass redeployment in online or in big clusters....

mfojtik commented:
@tnozicka i'm leaning toward to just ignore comparing the name in config change trigger for now to avoid mass rollouts. I will create issue to track this in 3.8.

tnozicka commented:
/lgtm

mfojtik commented on pkg/apps/apis/apps/test/ok.go:
```diff
@@ -166,6 +166,7 @@ func OkPodTemplate() *kapi.PodTemplateSpec {
 			DNSPolicy:                     kapi.DNSClusterFirst,
 			TerminationGracePeriodSeconds: &one,
 			SchedulerName:                 kapi.DefaultSchedulerName,
+			SecurityContext:               &kapi.PodSecurityContext{},
```
@kargakis @deads2k @tnozicka honestly, I have no idea why this is needed.... I spend last hour or so trying to figure out why the TestHandle() is failing, but without this the HasLatestPodTemplate() will see a diff where this field is nil vs. defaulted...

Just want to double check, but the codec i'm passing down to HasLatestPodTemplate is the same as we passed to the original trigger (annotationCodec), so I have no idea what happened there...

mfojtik replied:
also let this serve as argument for removing the annotations with encoded deployment configs

tnozicka replied:
@mfojtik I am not sure about this, it has been nil even before no matter the codec in tests - it might be sign of something else going wrong

mfojtik replied:
@tnozicka it might well be that previous state was wrong :-) i checked the encoded DC in RC and they seems correct (in running server).

mfojtik replied:
so my theory is this:

1) Before we haven't decoded the DC nor we did the deepequal in main deployment config controller, because that was part of generic trigger. 
2) The SecurityContext{} is set to empty by default, see kube testing for pod_specs.

So I think this change is fine.

mfojtik commented:
@tnozicka and we are green \o/

jim-minter commented:
@mfojtik are there any opportunities to increase test coverage in this codebase as part of your PR, e.g. around the case of repeated redeployments?

mfojtik commented:
@jim-minter i will want to do some trigger tests as follow ups. The continuous rollouts was broken because of the new logic that this PR introduced. It is good that it was found via templateinstance, but we should really have unit tests for this.

tnozicka commented:
I am ok with follow up
/lgmt

tnozicka commented:
/test unit
/test extended_networking_minimal

deads2k commented:
/retest
